src/Streamly/Internal/Data/Stream/IsStream.hs

```diff
@@ -676,3678 +654,250 @@ import Prelude hiding
-chunksOf2 n action f m = D.fromStreamD $ D.groupsOf2 n action f (D.toStreamD m)
-
--- | @arraysOf n stream@ groups the elements in the input stream into arrays of
--- @n@ elements each.
---
--- Same as the following but may be more efficient:
---
--- > arraysOf n = S.chunksOf n (A.writeN n)
---
--- /Internal/
-{-# INLINE arraysOf #-}
-arraysOf :: (IsStream t, MonadIO m, Storable a)
-    => Int -> t m a -> t m (Array a)
-arraysOf n = chunksOf n (writeNUnsafe n)
-
--- XXX we can implement this by repeatedly applying the 'lrunFor' fold.
--- XXX add this example after fixing the serial stream rate control
--- >>> S.toList $ S.take 5 $ intervalsOf 1 FL.sum $ constRate 2 $ S.enumerateFrom 1
--- > [3,7,11,15,19]
---
--- | Group the input stream into windows of @n@ second each and then fold each
--- group using the provided fold function.
---
--- @since 0.7.0
-{-# INLINE intervalsOf #-}
-intervalsOf
-    :: (IsStream t, MonadAsync m)
-    => Double -> Fold m a b -> t m a -> t m b
-intervalsOf n f xs =
-    splitWithSuffix isNothing (FL.lcatMaybes f)
-        (interjectSuffix n (return Nothing) (Serial.map Just xs))
+fromCallback :: MonadAsync m => ((a -> m ()) -> m ()) -> SerialT m a
+fromCallback setCallback = concatM $ do
+    (callback, stream) <- D.newCallbackStream
+    setCallback callback
+    return stream
 
 ------------------------------------------------------------------------------
--- N-ary APIs
+-- Specialized folds
 ------------------------------------------------------------------------------
 
--- | @groupsBy cmp f $ S.fromList [a,b,c,...]@ assigns the element @a@ to the
--- first group, if @b \`cmp` a@ is 'True' then @b@ is also assigned to the same
--- group.  If @c \`cmp` a@ is 'True' then @c@ is also assigned to the same
--- group and so on. When the comparison fails a new group is started. Each
--- group is folded using the fold @f@ and the result of the fold is emitted in
--- the output stream.
+-- |
+-- > drainN n = drain . take n
+-- > drainN n = fold (Fold.ltake n Fold.drain)
 --
--- >>> S.toList $ S.groupsBy (>) FL.toList $ S.fromList [1,3,7,0,2,5]
--- > [[1,3,7],[0,2,5]]
+-- Run maximum up to @n@ iterations of a stream.
 --
 -- @since 0.7.0
-{-# INLINE groupsBy #-}
-groupsBy
-    :: (IsStream t, Monad m)
-    => (a -> a -> Bool)
-    -> Fold m a b
-    -> t m a
-    -> t m b
-groupsBy cmp f m = D.fromStreamD $ D.groupsBy cmp f (D.toStreamD m)
-
--- | Unlike @groupsBy@ this function performs a rolling comparison of two
--- successive elements in the input stream. @groupsByRolling cmp f $ S.fromList
--- [a,b,c,...]@ assigns the element @a@ to the first group, if @a \`cmp` b@ is
--- 'True' then @b@ is also assigned to the same group.  If @b \`cmp` c@ is
--- 'True' then @c@ is also assigned to the same group and so on. When the
--- comparison fails a new group is started. Each group is folded using the fold
--- @f@.
+{-# INLINE drainN #-}
+drainN :: Monad m => Int -> SerialT m a -> m ()
+drainN n = drain . take n
```

Comment:
Shouldn't this be under `Eliminate`?

Reply:
The functions that should be in a separate module but aren't is mainly because the function depends on another independent module, like a function in Eliminate depending on a function in Generate. Since I wanted all these modules to not depend on each other, I moved them into the IsStream module. Later while refactoring StreamD I moved any such functions into a common StreamD module

Reply:
The classification of which function goes into which module is based on existing module structure is commented (haddocks) in the export list.